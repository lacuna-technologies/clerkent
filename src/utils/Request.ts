import axios from 'axios'

const request = axios.create({
  timeout: 10_000,
})

export default request