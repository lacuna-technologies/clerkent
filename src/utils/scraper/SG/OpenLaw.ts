import Request from 'utils/Request'
import Logger from 'utils/Logger'
import Constants from 'utils/Constants'
import { browser, Downloads } from 'webextension-polyfill-ts'
import Helpers from 'utils/Helpers'

const SEARCH_API_URL = `https://api.lawnet.sg/lawnet/search-service/api/lawnetcore/search/supreme-court`
const CITATION_API_URL = `https://api.lawnet.sg/lawnet/search-service/api/lawnetcore/document/citation`
const PDF_DOWNLOAD_URL = `https://api.lawnet.sg/lawnet/search-service/api/lawnetcore/download/v1`
const defaultUserDevice = {
  isBot: false,
}
const defaultConfig = {
  info: {
    appId: `APP_LAWNET_ONE`,
    sessionId: null,
    userDevice: defaultUserDevice,
    userId: 0,
    userName: null,
  },
}
const defaultSearchConfig = {
  ...defaultConfig,
  data: {
    filters: null,
    includeDraft: `true`,
    orderBy: `relevancy-desc`,
    pageLength: 10,
    searchDatabase: `allOfLawnet`,
    searchQuery: ``,
    searchSource: null,
    start: 1,
  },
}
const defaultCitationConfig = {
  ...defaultConfig,
  data: {
    citation: ``,
  },
}
const defaultPDFConfig = {
  ...defaultConfig,
  data: {
    citation: ``,
    legalTermsUrl: `/support/legal/terms-of-use`,
    pageUrl: ``,
  },
}
const defaultRequestOptions = {
  headers: {
    'Auth-Type': `Public`,
  },
}

const getCaseByCitation = async (citation: string): Promise<Law.Case[]> => {
  try {
    const requestBody = {
      ...defaultCitationConfig,
      data: {
        ...defaultCitationConfig.data,
        citation,
      },
    }
    const { data } = await Request.post(
      CITATION_API_URL,
      requestBody,
      defaultRequestOptions,
    )
    const encodedCitation = citation.replaceAll(` `, `+`)
    const judgmentLink: Law.Link = {
      doctype: `Judgment`,
      filetype: `HTML`,
      url: `https://www.lawnet.com/openlaw/cases/citation/${encodedCitation}`,
    }
    const summaryLink: Law.Link = {
      doctype: `Summary`,
      filetype: `HTML`,
      url: `https://www.lawnet.com/openlaw/cases/citation/${encodedCitation}`,
    }
    const result: Law.Case = {
      citation: data.data.metadata[`NeutralCitation`][`NCit`],
      database: Constants.DATABASES.SG_openlaw,
      jurisdiction: Constants.JURISDICTIONS.SG.id,
      links: [
        summaryLink,
        judgmentLink,
      ],
      name: data.data.metadata[`CaseTitle`],
    }
    return [result]

  } catch (error) {
    Logger.error(error)
    return []
  }
}

const getSearchResults = async (query: string): Promise<Law.Case[]>=> {
  const requestBody = {
    ...defaultSearchConfig,
    data: {
      ...defaultSearchConfig.data,
      searchQuery: query,
    },
  }
  try {
    const { data } = await Request.post(
      SEARCH_API_URL,
      requestBody,
      defaultRequestOptions,
    )
    const results: Law.Case[] = data.data.results.map((result): Law.Case => {
      const encodedCitation = result.ncitation.replaceAll(` `, `+`)
      const summaryLink: Law.Link = {
      doctype: `Summary`,
      filetype: `HTML`,
      url: `https://www.lawnet.com/openlaw/cases/citation/${encodedCitation}`,
    }
      const judgmentLink: Law.Link = {
        doctype: `Judgment`,
        filetype: `HTML`,
        url: `https://www.lawnet.com/openlaw/cases/citation/${encodedCitation}`,
      }
      return {
        citation: result.ncitation,
        database: Constants.DATABASES.SG_openlaw,
        jurisdiction: Constants.JURISDICTIONS.SG.id,
        links: [
          summaryLink,
          judgmentLink,
        ],
        name: result.titles[0],
      }
    })
    return results
  } catch (error){
    Logger.error(error)
    return []
  }
}

const getCaseByName = async (caseName: string): Promise<Law.Case[]> => {
  return await getSearchResults(caseName)
}

const getPDF = async (
  inputCase: Law.Case,
  inputDocumentType: Law.Link[`doctype`],
): Promise<void> => {
  Logger.log(`OpenLaw: getPDF`, inputCase, inputDocumentType)
  const options: Downloads.DownloadOptionsType = {
    body: JSON.stringify({
      ...defaultPDFConfig,
      data: {
        ...defaultPDFConfig.data,
        citation: inputCase.citation,
        pageUrl: (
          new URL(Helpers.getJudgmentLink(inputCase.links).url)
        ).pathname,
      },
    }),
    filename: Helpers.getFileName(inputCase, inputDocumentType),
    headers: [
      ...Object.entries(defaultRequestOptions.headers).map(([key, value]) => ({
        name: key,
        value,
      })),
      {
        name: `Content-Type`,
        value: `application/json`,
      },
    ],
    method: `POST`,
    saveAs: true,
    url: PDF_DOWNLOAD_URL,
  }
  Logger.log(options)
  await browser.downloads.download(options)
}

const OpenLaw = {
  getCaseByCitation,
  getCaseByName,
  getPDF,
  getSearchResults,
}

export default OpenLaw