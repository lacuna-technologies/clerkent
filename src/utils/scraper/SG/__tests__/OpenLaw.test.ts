import fs from 'fs'
import path from 'path'
import Request from 'utils/Request'
import OpenLaw from '../OpenLaw'

jest.mock(`utils/Request`)
const RequestPost = Request.post as jest.Mock

const CITATION = `[1999] SGHC 283`

describe(`SG OpenLaw`, () => {
  it(`should send valid request for getCaseByCitation`, () => {
    RequestPost.mockImplementationOnce(() => Promise.resolve({
      data: ``,
    }))
    OpenLaw.getCaseByCitation(CITATION)
    expect(RequestPost.mock.calls).toMatchSnapshot()
  })
  it(`should parse getCaseByCitation result correctly`, async () => {
    const mockResponse = JSON.parse(fs.readFileSync(
      path.join(__dirname, `responses`, `OpenLaw`, `getCaseByCitation-result.txt`),
    ).toString())
    RequestPost.mockImplementationOnce(() => Promise.resolve({ data: mockResponse }))
    const result = await OpenLaw.getCaseByCitation(CITATION)
    expect(result).toMatchSnapshot()
  })
})