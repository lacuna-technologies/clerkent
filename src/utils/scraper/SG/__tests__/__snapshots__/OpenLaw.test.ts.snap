// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SG OpenLaw should parse getCaseByCitation result correctly 1`] = `
Array [
  Object {
    "citation": "[1999] SGHC 283",
    "database": Object {
      "icon": "",
      "id": "SG_openlaw",
      "name": "OpenLaw",
      "url": "https://www.lawnet.com/openlaw",
    },
    "jurisdiction": "SG",
    "links": Array [
      Object {
        "doctype": "Summary",
        "filetype": "HTML",
        "url": "https://www.lawnet.com/openlaw/cases/citation/[1999]+SGHC+283",
      },
      Object {
        "doctype": "Judgment",
        "filetype": "HTML",
        "url": "https://www.lawnet.com/openlaw/cases/citation/[1999]+SGHC+283",
      },
    ],
    "name": "NV Sumatra Tobacco Co v Nanyang Brothers Tobacco Co Ltd",
  },
]
`;

exports[`SG OpenLaw should send valid request for getCaseByCitation 1`] = `
Array [
  Array [
    "https://api.lawnet.sg/lawnet/search-service/api/lawnetcore/document/citation",
    Object {
      "data": Object {
        "citation": "[1999] SGHC 283",
      },
      "info": Object {
        "appId": "APP_LAWNET_ONE",
        "sessionId": null,
        "userDevice": Object {
          "isBot": false,
        },
        "userId": 0,
        "userName": null,
      },
    },
    Object {
      "headers": Object {
        "Auth-Type": "Public",
      },
    },
  ],
]
`;
